mysql-test/suite/galera/t/galera_retry_applying.test

```diff
@@ -0,0 +1,103 @@
+#
+# Test retrying applying of a transaction
+#
+
+--source include/galera_cluster.inc
+
+#
+# Case 1: Retrying succeeds after one retry event, no error is raised.
+#
+CREATE TABLE t1 (f1 INTEGER PRIMARY KEY DEFAULT 0, f2 char(12));
+CREATE TABLE t3 (f1 INTEGER PRIMARY KEY DEFAULT 0, f2 char(12));
+
+START TRANSACTION;
+INSERT INTO t3 (f1, f2) VALUES (1, 'a');
+
+--connection node_2
+SET GLOBAL wsrep_applier_retry_count = 2;
+SET GLOBAL wsrep_innodb_insert_fail_table = "test/t1";
+SET GLOBAL debug_dbug = "+d,innodb_insert_fail:o,/dev/null";
+CALL mtr.add_suppression("Event .* Write_rows.* apply failed");
+
+--connection node_1
+INSERT INTO t3 (f1, f2) VALUES (2, 'b');
+INSERT INTO t1 (f1, f2) VALUES (3, 'c'), (4, 'd'), (5, 'e');
+INSERT INTO t3 (f1, f2) VALUES (3, 'c');
+INSERT INTO t1 (f1, f2) VALUES (1, 'a'), (2, 'b');
+COMMIT;
+
+--connection node_1
+SELECT COUNT(*) AS expect_5 FROM t1;
+SELECT COUNT(*) AS expect_3 FROM t3;
+
+--connection node_2
+SELECT COUNT(*) AS expect_5 FROM t1;
+SELECT COUNT(*) AS expect_3 FROM t3;
+
+#
+# Cleanup after Case 1.
+#
+
+--connection node_1
+SET GLOBAL wsrep_applier_retry_count = 0;
+SET DEBUG_SYNC = 'RESET';
+SET GLOBAL debug_dbug = DEFAULT;
+DROP TABLE t1;
+DROP TABLE t3;
+
+
+#
+# Case 2: Slave retries applying of a transaction multiple times. All
+# retry attempts fail, and the applying will fail with the expected
+# error.
+#
+
+CREATE TABLE t2 (f1 INTEGER PRIMARY KEY DEFAULT 0, f2 char(12));
+
+--connection node_2
+CALL mtr.add_suppression("Event .* Write_rows.* apply failed");
+CALL mtr.add_suppression("Inconsistency detected");
+CALL mtr.add_suppression("Failed to apply write set");
+
+SET GLOBAL wsrep_applier_retry_count = 2;
+SET GLOBAL wsrep_innodb_insert_fail_table = "test/t1";
+SET GLOBAL debug_dbug = "+d,innodb_insert_fail:o,/dev/null";
+
+--connection node_1
+START TRANSACTION;
+INSERT INTO t2 (f1, f2) VALUES (1, 'a'), (2, 'b');
+COMMIT;
+
+# shutdown node 2
+--connection node_2
+--echo Shutting down server ...
+SET wsrep_on=OFF;
+--source include/shutdown_mysqld.inc
+--remove_file $MYSQLTEST_VARDIR/mysqld.2/data/grastate.dat
```

Comment:
here is a race condition: previously a transaction was committed in node_1 and here node_2 will shutdown. But there is no check for the fate of the replicated INSERT from node_1: it may still be replicating or is currently applying or has already committed in node_2. For deterministic test behavior, the state of the INSERT transaction should be synced here or documented if it does not matter for the test result and can be safely ignored

Reply:
also, as this test phase is supposed to cause sure applier failure, the node should crash, so no need to shutdown it anymore.

Reply:
Removed table names in _row0ins.c_, but `DBUG_EXECUTE_IF` macros still
mention the test database _test_. There are now two separate `DBUG_EXECUTE_IF` labels: 
  `innodb_insert_fail_once`  -  for failing INSERT once inside InnoDB, and
  `innodb_insert_fail_always`  -  for failing INSERT always inside InnoDB.

Synchronization.
There are now 4 synchronization points in the MTR test, two in each of
the 2 test cases:

1.  Test case 1: wait till the insert transaction has been replicated and committed in node_2 (line 25)

2.  Test case 1: wait till the transaction has been replicated and committed in node_2 (line 44)

3.  Test case 2: wait for node_2 to crash (line 54)
 
4.  Test case 2: wait till node 2 is back in the cluster (line 120) 

The test does not work without shutting down the server on node 2 after applier failure.